Doc reorg: Configuring CumulusCI (#2117)
SFDO-Tooling/CumulusCI: docs/cookbook.rst

Cookbook
========

Task Recipes
------------

Run a Shell Command
******************************

.. code-block:: yaml

    run_custom_command:
        description: Greets the user
        class_path: cumulusci.cli
        options: 
            command: "echo 'Hello there!"

        

Run a ``sfdx`` Command
****************************
The ``dx`` task lets you run an arbitrary ``sfdx`` command.
You can perform this with ``cci`` on a terminal::

    $ cci task run dx -o command 'force:api:limits:display'

Or you can utilize the same ``class_path`` as the ``dx`` task and make a custom task that can be executed by itself or as a step in a flow.

.. code-block:: yaml

    dx_limits:
        description: Display
        class_path: cumulusci.tasks.sfdx.SFDXBaseTask
        group: dx 
        options: 
            command: sfdx force:limits:api:display

In this case, we actually utilize ``SFDXBaseTask``, if you would like to run a ``sfdx`` command that references an org, utilize ``SFDXOrgTask`` instead.

Custom Deploy
************************

> It is often useful to be able to define multiple custom deploy tasks to be able to easily identify which task deploys a specific subsection of Metadata.

Reviewer: ```suggestion
It is often useful to be able to define multiple custom deployment tasks, particularly when working with <TODO: link Unpackaged Metadata>.
```

> For example,  here is a custom task that is defined to only deploy reports::

Reviewer: This might be a little confusing - it doesn't do any _filtering_ of metadata.

Cookbook
========

Task Recipes
------------

Run a Shell Command
******************************

.. code-block:: yaml

    run_custom_command:
        description: Greets the user
        class_path: cumulusci.cli
        options: 
            command: "echo 'Hello there!"

        

Run a ``sfdx`` Command
****************************
The ``dx`` task lets you run an arbitrary ``sfdx`` command.

> You can perform this with ``cci`` on a terminal::

Reviewer 2: Combine and reword to avoid using "this" without a subject.

    $ cci task run dx -o command 'force:api:limits:display'

> Or you can utilize the same ``class_path`` as the ``dx`` task and make a custom task that can be executed by itself or as a step in a flow.

Reviewer 2: ```suggestion
Or use the same ``class_path`` as the ``dx`` task and make a custom task that can be executed by itself or as a step in a flow.
```
https://confluence.internal.salesforce.com/display/DOCTEAM/utilize

.. code-block:: yaml

    dx_limits:
        description: Display
        class_path: cumulusci.tasks.sfdx.SFDXBaseTask
        group: dx 
        options: 
            command: sfdx force:limits:api:display

> In this case, we actually utilize ``SFDXBaseTask``, if you would like to run a ``sfdx`` command that references an org, utilize ``SFDXOrgTask`` instead.

Reviewer 3: We should use SFDXOrgTask for this example. It doesn't make sense to run the force:limits:api:display command without an org.

> In this case, we actually utilize ``SFDXBaseTask``, if you would like to run a ``sfdx`` command that references an org, utilize ``SFDXOrgTask`` instead.

Reviewer 2: ```suggestion
In this case, we actually use ``SFDXBaseTask``. To run an ``sfdx`` command that references an org, use ``SFDXOrgTask`` instead.
```
Not sure if this edit still applies given David's comment. But the gist is to break up the run-on sentence.

Custom Deploy
************************

> It is often useful to be able to define multiple custom deployment tasks that deployg a specific subset of your projects metadata.

Reviewer 2: ```suggestion
It's often useful to define multiple custom deployment tasks that deploy a specific subset of your project's metadata.
```

> This is particularly true when working with `unpackaged Metadata <TODO>`_.

Reviewer 2: Combine and reword to avoid using "this" without a subject.

Reviewer: Missing possessive apostrophe in `projects` also.

> Here is a custom task that is defined to only deploy only the metadata contained in ``unmanaged/config/reports``.

Reviewer 2: ```suggestion
This custom task is defined to deploy only the metadata contained in ``unmanaged/config/reports``.
```

.. code-block:: yaml

    deploy_reports:
        description: Deploy Reports 
        class_path: cumulusci.tasks.salesforce.Deploy
        options:
            path: unmanaged/config/reports    

> Being able to give this task a new name makes it much more intuitive as to what the task is actually doing.

Reviewer 2: ```suggestion
Giving this task a descriptive name makes the task's purpose much more intuitive.
```
In sales, they say "Always be selling." In doc, we say "Always be trimming." This sentence is a prime candidate for tightening up for clarity and readability.

> Multiple custom deploy tasks like this allow NPSP to `create flows <https://github.com/SalesforceFoundation/NPSP/blob/87daa94f9494d28ce3a5cc52bd5d5308cc804a2b/cumulusci.yml#L692>` that make it easy to define the order that Metadata is deployed in.

Reviewer 2: ```suggestion
Multiple custom deploy tasks such as this one let NPSP `create flows <https://github.com/SalesforceFoundation/NPSP/blob/87daa94f9494d28ce3a5cc52bd5d5308cc804a2b/cumulusci.yml#L692>` to define the order for deploying metadata.
```

            

Task to Execute Anonymous Apex
*********************************

> The following shows an example task named ``project_default_settings`` which runs the public static method ``initializeProjectDefaults()`` located in file ``scripts.initialize.cls``::

Reviewer 2: ```suggestion
This sample task named ``project_default_settings`` runs the public static method ``initializeProjectDefaults()`` located in the file ``scripts.initialize.cls``::
```

Reviewer 2: Consider providing a bit more context about when to run it and how it relates to other discussion about defaults.
